core/txindexer.go

```diff
@@ -0,0 +1,215 @@
+type txIndexer struct {
+	// limit is the maximum number of blocks from head whose tx indexes
+	// are reserved:
+	//  * 0: means the entire chain should be indexed
+	//  * N: means the latest N blocks [HEAD-N+1, HEAD] should be indexed
+	//       and all others shouldn't.
+	limit    uint64
+	db       ethdb.Database
+	progress chan chan TxIndexProgress
+	term     chan chan struct{}
+	closed   chan struct{}
+
+	chain *BlockChain
+}
+
+// newTxIndexer initializes the transaction indexer.
+func newTxIndexer(limit uint64, chain *BlockChain) *txIndexer {
+	indexer := &txIndexer{
+		limit:    limit,
+		db:       chain.db,
+		progress: make(chan chan TxIndexProgress),
+		term:     make(chan chan struct{}),
+		closed:   make(chan struct{}),
+		chain:    chain,
+	}
+	chain.wg.Add(1)
+	go func() {
+		defer chain.wg.Done()
+		indexer.loop(chain)
+	}()
+
+	var msg string
+	if limit == 0 {
+		msg = "entire chain"
+	} else {
+		msg = fmt.Sprintf("last %d blocks", limit)
+	}
+	log.Info("Initialized transaction indexer", "range", msg)
+
+	return indexer
+}
+
+// run executes the scheduled indexing/unindexing task in a separate thread.
+// If the stop channel is closed, the task should be terminated as soon as
+// possible, the done channel will be closed once the task is finished.
+func (indexer *txIndexer) run(tail *uint64, head uint64, stop chan struct{}, done chan struct{}) {
+	start := time.Now()
+	indexer.chain.txIndexTailLock.Lock()
+	defer func() {
+		txUnindexTimer.Inc(time.Since(start).Milliseconds())
+		indexer.chain.txIndexTailLock.Unlock()
+		close(done)
+		indexer.chain.wg.Done()
+	}()
+
+	// Short circuit if chain is empty and nothing to index.
+	if head == 0 {
+		return
+	}
+
+	// Defensively ensure tail is not nil.
+	if tail == nil {
+		tail = new(uint64)
+	}
+
+	if head-indexer.limit+1 >= *tail {
+		// Unindex a part of stale indices and forward index tail to HEAD-limit
+		rawdb.UnindexTransactions(indexer.db, *tail, head-indexer.limit+1, stop, false)
+	}
+}
+
+// loop is the scheduler of the indexer, assigning indexing/unindexing tasks depending
+// on the received chain event.
+func (indexer *txIndexer) loop(chain *BlockChain) {
+	defer close(indexer.closed)
+
+	// If the user just upgraded to a new version which supports transaction
+	// index pruning, write the new tail and remove anything older.
+	if rawdb.ReadTxIndexTail(indexer.db) == nil {
+		rawdb.WriteTxIndexTail(indexer.db, 0)
+	}
+
+	// Listening to chain events and manipulate the transaction indexes.
+	var (
+		stop     chan struct{}                       // Non-nil if background routine is active.
+		done     chan struct{}                       // Non-nil if background routine is active.
+		lastHead uint64                              // The latest announced chain head (whose tx indexes are assumed created)
+		lastTail = rawdb.ReadTxIndexTail(indexer.db) // The oldest indexed block, nil means nothing indexed
+
+		headCh = make(chan ChainEvent, 1)
```

Comment:
we got a response, seems we don't need a buffered ch: https://github.com/ethereum/go-ethereum/pull/30350

I totally forgot we run the indexer with another gouroutine, so should be good to use a unbuffered ch.

Reply:
let's change it to an unbuffered ch to keep it aligned with upstream. sorry for changing this twice.